drivers/net/native_sim_adapt.c

```diff
@@ -0,0 +1,648 @@
+{
+	struct sockaddr_storage addr_storage;
+	struct sockaddr *addr = (struct sockaddr *)&addr_storage;
+	socklen_t addrlen;
+	int ret;
+
+	ret = sockaddr_from_nsim(&addr, &addrlen, nsim_addr, nsim_addrlen);
+	if (ret < 0) {
+		return ret;
+	}
+
+	ret = bind(fd, addr, addrlen);
+	if (ret < 0) {
+		return -errno_to_nsim(errno);
+	}
+
+	return ret;
+}
+
+int nsim_adapt_connect(int fd, const struct nsim_sockaddr *nsim_addr, size_t nsim_addrlen)
+{
+	struct sockaddr_storage addr_storage;
+	struct sockaddr *addr = (struct sockaddr *)&addr_storage;
+	socklen_t addrlen;
+	int ret;
+
+	ret = sockaddr_from_nsim(&addr, &addrlen, nsim_addr, nsim_addrlen);
+	if (ret < 0) {
+		return ret;
+	}
+
+	ret = connect(fd, addr, addrlen);
+	if (ret < 0) {
+		return -errno_to_nsim(errno);
+	}
+
+	return ret;
+}
+
+int nsim_adapt_listen(int fd, int backlog)
+{
+	int ret;
+
+	ret = listen(fd, backlog);
+	if (ret < 0) {
+		return -errno_to_nsim(errno);
+	}
+
+	return ret;
+}
+
+int nsim_adapt_accept(int fd, struct nsim_sockaddr *nsim_addr, size_t *nsim_addrlen)
+{
+	struct sockaddr_storage addr_storage;
+	struct sockaddr *addr = (struct sockaddr *)&addr_storage;
+	socklen_t addrlen;
+	int ret;
+
+	ret = sockaddr_from_nsim(&addr, &addrlen, nsim_addr, *nsim_addrlen);
+	if (ret < 0) {
+		return ret;
+	}
+
+	ret = accept(fd, addr, &addrlen);
+	if (ret < 0) {
+		return -errno_to_nsim(errno);
+	}
+
+	sockaddr_to_nsim(addr, addrlen, nsim_addr, nsim_addrlen);
+
+	return ret;
+}
+
+int nsim_adapt_sendto(int fd, const void *buf, size_t len, int flags,
+		      const struct nsim_sockaddr *nsim_addr, size_t nsim_addrlen)
+{
+	struct sockaddr_storage addr_storage;
+	struct sockaddr *addr = (struct sockaddr *)&addr_storage;
+	socklen_t addrlen;
+	int ret;
+
+	ret = sockaddr_from_nsim(&addr, &addrlen, nsim_addr, nsim_addrlen);
+	if (ret < 0) {
+		return ret;
+	}
+
+	ret = sendto(fd, buf, len,
+		     socket_flags_from_nsim(flags) | MSG_NOSIGNAL,
+		     addr, addrlen);
+	if (ret < 0) {
+		return -errno_to_nsim(errno);
+	}
+
+	return ret;
+}
+
+int nsim_adapt_recvfrom(int fd, void *buf, size_t len, int flags,
+			struct nsim_sockaddr *nsim_addr, size_t *nsim_addrlen)
+{
+	struct sockaddr_storage addr_storage;
+	struct sockaddr *addr = (struct sockaddr *)&addr_storage;
+	socklen_t addrlen;
+	int ret;
+
+	ret = sockaddr_from_nsim(&addr, &addrlen, nsim_addr, *nsim_addrlen);
+	if (ret < 0) {
+		return ret;
+	}
+
+	ret = recvfrom(fd, buf, len, socket_flags_from_nsim(flags),
+		       addr, &addrlen);
+	if (ret < 0) {
+		return -errno_to_nsim(errno);
+	}
+
+	sockaddr_to_nsim(addr, addrlen, nsim_addr, nsim_addrlen);
+
+	return ret;
+}
+
+int nsim_adapt_poll(struct nsim_pollfd *fds, int nfds, int timeout)
+{
+	struct pollfd host_fds[nfds];
+	int ret;
+
+	for (int i = 0; i < nfds; i++) {
+		host_fds[i].fd = fds[i].fd;
+		host_fds[i].events = fds[i].events;
+		host_fds[i].revents = fds[i].revents;
+	}
+
+	ret = poll(host_fds, nfds, timeout);
```

Comment:
This is not going to work as you expect if there is any timeout. (You are going to hang the whole zephyr until the call to linux's poll() returns)

I presume if there is a timeout and it is not ready, you'd want it to k_yield().
So, this would need to be replaced with a while loop (in the caller), where this calls the linux poll() with a timeout of 0, and (in the caller) while there is time left and the descriptor is not ready (ret==0), you k_sleep() for some kconfigured amount of time, before calling this function to do a the linux poll() again.

Reply:
You were right. i have implemented your suggestion with `k_sleep()` on the Zephyr (caller) in the first place. However, then I reworked the code to utilize `NSI_HW_EVENT()`, which seems the perfect mechanism for polling host interfaces. The advantage of the latter is that there is minimal latency between traffic on the host and notification of Zephyr. This makes it much more performant (better network transfer speeds). Additionally it makes it independent of systick/counter/timer configuration, which affects `k_sleep()` behavior and its latency.